### DISCUSSION TITLE
Whole Body Simulator - Add Custom Robots

### DISCUSSION THREAD

**Discussion by nunoguedelha** [ami-iit/matlab-whole-body-simulator#92] [Q&A]
Quoting below the question submitted by @oztahmet :

> Hello Nuno,
> 
> I hope you are well.
> 
> I'm using Whole Body Simulator for my graduate project. A demo program called "torqueControlBalancingWithSimu" is running well. However, I'd like to add a custom fixed-point multi-DOF manipulator. How can I achieve this?
> 
> It seems URDF format and meshes are suitable for adding robots, and I can edit configuration files as well. However, I'm unsure where to place these files and how to compile/rebuild the library (if necessary). I apologize, but I'm not familiar with binary or source installations using conda.
> 
> The library was installed using a single command for MATLAB on Windows.
> 
> Ahmet,
> ITU - Control and Automation Dept.

**Comment by nunoguedelha** [comment]
Hello @oztahmet,

I'm glad you get to try this simulation framework, and any further feedback on your experience with following the installation instructions and using the software is welcomed.

Regarding your questions:
- Yes, you should be able to add a custom robot with your own URDF and meshes files (although it's less trivial than just selecting a model among the default ones).
    You can check where the simulator looks for the URDF file by running:
    ```
    system('yarp resource --find model.urdf')
    ```
- The simulator framework indeed supports fixed-base as well as floating-base models.
    As far as I remember, you should set the variable `robot_config.robotFrames.FIXED`, in the config files, to the name of the frame you want fixed.

I haven't worked on this framework for while now, so I need to reproduce your use-case and check some details and get back to you in a few, unless @traversaro already has additional information or clarifications to add.

**Reply by oztahmet** [reply]
Hello @nunoguedelha,

Thanks for the quick response. I tried to implement a control loop using the RRbot example. I have two questions regarding this topic.

- [ ] 1) Fixed the base frame using `robot_config.robotFrames.FIXED`, but there is a continuous downward movement along the z-axis during the simulation. Irrlicht visualization is used. My expectation is for the robot base to remain fixed to the ground. I kept the friction coefficient (fric_coeff = 10) and footprint settings (a bit larger footprint) the same in the config file settings.

- [x] 2) When I stop the simulation in Simulink, change and reinitialize the robot settings from the config file, and use the new config variables in the Matlab workspace to start the simulation again, the **OpenMP** forces Matlab to close. However, when I use the same robot settings in the Matlab workspace and restart the simulation several times, it works well. 

> **Error code:** 
> _MP: Error #15: Initializing libomp.dll, but found libiomp5md.dll already initialized.
> OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program. That is dangerous, since it can degrade performance or cause incorrect results. The best thing to do is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding static linking of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented workaround you can set the environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, but that may cause crashes or silently produce incorrect results. For more information, please see http://openmp.llvm.org/_

### edit: 2nd issue is not related with Irrlicht engine, its more general with libraries. The reference comment to the Irrlicht engine was removed.

**Reply by oztahmet** [reply]
For the 2nd issue,

> MP: Error https://github.com/ami-iit/matlab-whole-body-simulator/pull/15: Initializing libomp.dll, but found libiomp5md.dll already initialized.

Search the  `libiomp5md.dll` files in the  `robotology-matlab` folder and delete them. This solution worked for me!

**Reply by traversaro** [reply]
For the second error, a simple fix is to use the `netlib` blas library, see https://github.com/robotology/robotology-superbuild/blob/v2024.02.0/doc/faqs.md#how-to-i-solved-the-initializing-libompdylib-but-found-libiomp5dylib-already-initialized-when-using-matlab-libraries . I will update the one-line installer to use that info, thanks for reporting the problem @oztahmet !

**Reply by traversaro** [reply]
See https://github.com/robotology/robotology-superbuild/pull/1658 .

**Reply by oztahmet** [reply]
@traversaro @nunoguedelha,

For the 1st issue,

https://github.com/ami-iit/matlab-whole-body-simulator/issues/18#issuecomment-804023901

I can't figure out how to fix base point to the ground. It just sinking with respect to friction coefficient magnitude.

![image](https://github.com/ami-iit/matlab-whole-body-simulator/assets/19744831/717707ab-dd77-46c8-b3b7-6ebf5c4f7951)

**Reply by nunoguedelha** [reply]
Hello @oztahmet ,
I was having issues installing the simulator and running a simple example on my machine (MacOS M1) because of the unavailability of `robotology` conda packages for Apple Silicon platforms (https://github.com/ami-iit/matlab-whole-body-simulator/discussions/93). While @traversaro is working on a proper fix, I worked around the issue on my setup and could reproduce the issue.

I just started analysing it but I would already say:
- If I remember correctly, the fixed-base models (and variable `robot_config.robotFrames.FIXED`) are properly supported only by the closed-chain simulator `RobotDynWithContacts_ClosedChain`:
    ![image](https://github.com/ami-iit/matlab-whole-body-simulator/assets/6848872/8c7c880f-b7e0-4d41-9ade-d9a8904936dd)
    ...and I guess you are using the other block, meant for systems without closed chains `RobotDynWithContacts`.
- As far as I remember, the RRBot model was created before the implementation of the support of fixed-base models. So RRbot is actually a floating-base model which uses a workaround which consists in having a very massive base with huge terms in the inertial matrix and significant width. This property, along with the usual handling of floating-base systems and ground impacts, was enough to keep the base "fixed".
- Considering for now the use of `RobotDynWithContacts`, I suspect there is an issue with the ground impact detection when the initial position xyz of the base is exactly `[0, 0, 0]`.
=> WORKAROUND: you can already use a workaround by setting the initial z position just slightly above the ground, e.g. `0.001 m`. You can set it in the file `app/robots/RRbot1/configRobot.m`:
```
% Initial condition of iCub and for the integrators.
initialConditions.base_position = [0; 0; 0.001];
```

Meanwhile, we will analyse the issue and check the use of `RobotDynWithContacts_ClosedChain `.

**Reply by nunoguedelha** [reply]
## On my previously suggested workaround...
...adding just a plot of the z coordinate time series, for different initial heights `z = 0.001`, `z = 0` and `z = -0.001` (in metres), illustrating the effect of the workaround I suggested in my previous post:

![image](https://github.com/ami-iit/matlab-whole-body-simulator/assets/6848872/934f69fd-3872-4807-b166-1dda14c261cf)

## Joint Velocity STOPS Being Updated
I ran a test while forcing the input joint torques to 0 and initial position of the robot (double pendulum) at horizontal position (q1 = 90°, q2 = 0º) as we can see depicted below:

![image](https://github.com/ami-iit/matlab-whole-body-simulator/assets/6848872/2eeeb5ff-931c-4a92-9c26-bc66b35a7448)

As the simulation runs, we see the base lying on the ground and the pendulum still in the initial position while the torques are null, which is physically wrong and indicates that the model state is not updated properly. Indeed the plots show that the velocity stops being updated as soon as the base hits the ground (`z <w_H_b>(3,4) = 0`).

![image](https://github.com/ami-iit/matlab-whole-body-simulator/assets/6848872/dc8936df-edb5-4549-bb24-d7480a5ef46d)
Zoom in...
![image](https://github.com/ami-iit/matlab-whole-body-simulator/assets/6848872/0dc9ab00-4310-489b-b45a-07627b1efd76)

## Test with the Closed-Chain Simulator

Ongoing...

**Comment by nunoguedelha** [comment]
## Test with the Closed-Chain Simulator

Created Simulink model for testing the library RobotDynWithContacts_ClosedChain on iCubGenova04 or RRbot1 models.
Copied base configuration and model files from this repository's root path:
- [initFcn.m](https://github.com/ami-iit/matlab-whole-body-simulator/blob/89ad53de8a82f16180ef12dc2c8b4b3d68990ecb/initFcn.m)
- [postLoadFcn.m](https://github.com/ami-iit/matlab-whole-body-simulator/blob/89ad53de8a82f16180ef12dc2c8b4b3d68990ecb/postLoadFcn.m)
- [./app](https://github.com/ami-iit/matlab-whole-body-simulator/tree/89ad53de8a82f16180ef12dc2c8b4b3d68990ecb/app)
- [./+src](https://github.com/ami-iit/matlab-whole-body-simulator/tree/89ad53de8a82f16180ef12dc2c8b4b3d68990ecb/%2Bsrc)
- [./test_matlab_system_2020b.mdl](https://github.com/ami-iit/matlab-whole-body-simulator/blob/89ad53de8a82f16180ef12dc2c8b4b3d68990ecb/test_matlab_system_2020b.mdl) -> changes saved to `test_matlab_system_closed_chain_2024b.slx` to be pushed to a branch.


### Running the model with free-falling iCubGenova04

- Set `l_sole` and `r_sole` frames as `IN_CONTACT_WITH_GROUND` robot frames instead of `LEFT_FOOT` and `RIGHT_FOOT`.
- Fixed the computation of total number of frames in contact with the ground as it was hard-coded. This is important if you define less or more than two frames which can get in contact with the ground.

Final Robot frames list is:
```
Frames.BASE = 'root_link';
Frames.IMU = 'head_imu_0';
Frames.COM = 'com';
Frames.IN_CONTACT_WITH_GROUND = {'l_sole','r_sole'};
robot_config.robotFrames = Frames;

contact_config.foot_print = vertex;
contact_config.total_num_vertices = size(vertex,2)*length(Frames.IN_CONTACT_WITH_GROUND);
```

The simulation runs successfully


### Running the model with iCubGenova04 on a pole (invisible)

For this test we set a fixed frame:
- We set the robot base a bit higher: `initialConditions.base_position = [0; 0; 1];`.
- We set the fixed frame: `Frames.FIXED = Frames.BASE;`.
- We leave `Frames.IN_CONTACT_WITH_GROUND` unchanged. This can be useful if for instance the pole is not high enough and the feet touch the ground at some point of the simulation.

In addition we integrate a simple gravity compensation in the controller.

The simulation runs successfully (almost).

We observe the robot (base) shifting downwards at constant speed, as we saw happening previously with the pendulum model with the simulator library `RobotDynWithContacts` (no closed chains). Not sure both problems are rellated.

### Running the model with RRbot1 fixed on the ground

For this test, we set RRbot1 configuration to work with `RobotDynWithContacts_ClosedChain`:
- The base frame of the double pendulum is sitting on the ground:
  ```
  initialConditions.base_position = [0; 0; 0];
  Frames.FIXED = Frames.BASE;
  ````
- Set last link's frame (end effector) to be the IN_CONTACT_WITH_GROUND frame and add four vertices to it:
  ```
  Frames.IN_CONTACT_WITH_GROUND = {'RRbot_hokuyo_link'};
  contact_config.friction_coefficient = 0.1;
  vertex = zeros(3, 4);
  vertex(:, 1) = [-0.01; +0.01; 0];
  vertex(:, 2) = [+0.01; +0.01; 0];
  vertex(:, 3) = [+0.01; -0.01; 0];
  vertex(:, 4) = [-0.01; -0.01; 0];
  ```
- the "world" link in the URDF model has no mass, as we don't need anymore a very large mass and angular inertia to emulate the ground:
  ```
  <link name="world"/>
  ```

The simulation runs successfully (almost).

We observe the robot leaning to the right and its base sink into the ground at constant speed. We can guess the observed drift is the same behaviour observed on the iCub model. Apparently, the constraint conditions used to keep the base fixed are not "strong" enough, or not well posed, causing it to drift downwards and topple. Typically, the constraint conditions could be ensuring zero acceleration of the "FIXED" frame, instead of zero velocity. We've seen this before in old simulations with iCub feet fixed through holonomic constraints (zero acceleration of the feet).
**==> TO BE ANALYZED**.

@traversaro , I think an issue could already be raised for this fixed frame drift problem.

**Reply by traversaro** [reply]
> @traversaro , I think an issue could already be raised for this fixed frame drift problem.

Ok, feel free to open it, thanks!

**Reply by nunoguedelha** [reply]
Submitted #94 .
